libafl_qemu/src/executor.rs

```diff
@@ -314,7 +318,9 @@ where
         fuzzer: &mut Z,
         state: &mut S,
         event_mgr: &mut EM,
+        executor_hooks: HT,
```

Comment:
The executor should probably not hold the hooks. If the executor is ever wrapped, it will make it very difficult to write code for (i.e. it will require a custom implementation of `ExecutorHooksTuple` which will likely require unsafe, see: the implementation for accessing observers in `DifferentialExecutor`.

Moreover, there seem to be some hooks that are _required_ by certain executors. Why should the user be responsible for initialising these? It seems like a footgun (e.g. accidentally using `InChild` without forking?).

Instead, we should define associated types for the executor-required hooks (e.g. `InChildProcessHook` for `InProcessForkExecutor`) and make the `ExecutorHook` generic over `E`. Then we define initialisers which are generic over executors s.t. we concretise the hook's types for all of its methods. Then, we no longer need to have generic methods in the trait definition.

Reply:
> Moreover, there seem to be some hooks that are required by certain executors. Why should the user be responsible for initialising these? It seems like a footgun (e.g. accidentally using InChild without forking?).

for this issue, it is easy to fix. 
we can have another method `with_hooks` as the constructor of the executor.
Then, in the constructor we just add the "required" hooks, and merge it like
```
tuple_list!(default_hooks, <any other stuff from user>)
```
the point is to make it configurable, and make it possible to add any hooks (not just the default hook) later on.

> Instead, we should define associated types for the executor-required hooks (e.g. InChildProcessHook for InProcessForkExecutor) and make the ExecutorHook generic over E. Then we define initialisers which are generic over executors s.t. we concretise the hook's types for all of its methods. Then, we no longer need to have generic methods in the trait definition.

Do you mean trait ExecutorHooks has should have the associate type E: Executor?

Reply:
> We could add and remove hooks at runtime, too

yes is ideal but require dyn dispatch though.

Reply:
I still don't think dynamic dispatch should be necessary or desirable here. If the user wants to do something dynamically, they can register a hook that simply handles the dynamic callbacks; it shouldn't be necessary to dynamically type it.

> Do you mean trait ExecutorHooks has should have the associate type E: Executor?

No, I mean to define `trait ExecutorHook` so that the hook a) may only be defined for certain executors, rather than requiring support for all executors as it is now, and b) ensures that the implementation of the trait is fixed so we can _potentially_ use dynamic dispatch in the future (dynamic dispatch is not possible over generic methods).

Reply:
```
I mean to define trait ExecutorHook<E>
```
I can't do this because the Executor holds the Hooks, like `HT: ExecutorHook<Self>` then the compiler can't evaluate the type requirement due to infinite recursion

Reply:
See above:
> The executor should probably not hold the hooks. If the executor is ever wrapped, it will make it very difficult to write code for (i.e. it will require a custom implementation of `ExecutorHooksTuple` which will likely require unsafe, see: the implementation for accessing observers in `DifferentialExecutor`).

Maybe the fuzzer should hold this?

Reply:
that's a problem when two different executors are used in stage_tuples..
The point is to make each executor has customizable hooks like the current QemuExecutor.